binary_search_tree/binary_search_tree.py

```diff
@@ -1,17 +1,82 @@
 class BinarySearchTree:
-  def __init__(self, value):
-    self.value = value
-    self.left = None
-    self.right = None
+    def __init__(self, value):
+        self.value = value
+        self.left = None
+        self.right = None
 
-  def insert(self, value):
-    pass
+    def insert(self, value):
+        # Dont forget to wrap the value in a node
 
-  def contains(self, target):
-    pass
+        # 1. Compare the element againt current value
+        # 2. Based on the result of the comparison go left or right
+        # 3. If we find an empty spot park element there
+        # 4. Otherwise go back to step 1
 
-  def get_max(self):
-    pass
+        # Base case?
+        # base: we have found an empty spot where we can add the value
 
-  def for_each(self, cb):
-    pass
\ No newline at end of file
+        if value < self.value:
+            # if value is less, we go left
+            # if there is no left child, we can put node here
+            if not self.left:
```

Comment:
Try combining the two nested if statements and see if you get the same result.
```
if value < self.value and not self.left:
    self.left = BinarySearchTree(value)
elif value < self.value and self.left:
    self.left.insert(value)
```
You can do the same with the other nested elif statement. Here is how I did mine...
```
    def for_each(self, cb):
        value = self.value
        left = self.left
        right = self.right
        if value is None:
            return cb(value)
        elif left is None and right is None:
            cb(value)
        if left is not None:
            self.left.for_each(cb)
            cb(value)
        if right is not None:
            self.right.for_each(cb)
            cb(value)
```